R/sample.R

```diff
@@ -34,47 +44,107 @@ mcstate_sample <- function(model, sampler, n_steps, initial = NULL) {
+                    "that many entries, each of which has {n_pars}",
+                    "element{?s}")),
+        arg = "initial", call = call)
+    }
+    err <- lengths(initial) != n_pars
+    if (any(err)) {
+      ## The format here is not beautiful but it's fairly informative
+      ## at least.
+      len <- as.character(lengths(initial)[err])
+      len[[1]] <- paste(len[[1]], ngettext(len[[1]], "value", "values"))
+      detail <- sprintf("%d (%s)", which(err), len)
+      cli::cli_abort(
+        c(paste("Unexpected initial parameter length; expected {n_pars}"),
+          i = "Incorrect length for element{?s} {detail}",
+          i = paste("Your model has {n_pars} parameter{?s}, so each",
+                    "element of your list must have this many elements")),
+        arg = "initial", call = call)
+    }
+    initial <- matrix(unlist(unname(initial)), n_chains, n_pars, byrow = TRUE)
+  } else if (is.matrix(initial)) {
+    hint_matrix <- paste(
+      "Expected a matrix with {n_chains} row{?s} (1 per chain) and",
+      "{n_pars} column{?s} (1 per parameter)")
+    if (nrow(initial) != n_chains) {
+      cli::cli_abort(
+        c(paste("Unexpected number of rows in 'initial'",
+                "(given {nrow(initial)}, expected {n_chains})"),
+          i = hint_matrix),
+        arg = "initial", call = call)
+    }
+    if (ncol(initial) != n_pars) {
+      cli::cli_abort(
+        c(paste("Unexpected number of columns in 'initial'",
+                "(given {ncol(initial)}, expected {n_pars})"),
+          i = hint_matrix),
+        arg = "initial", call = call)
+    }
   } else {
-    pars <- initial
-    if (length(pars) != length(model$parameters)) {
+    if (length(initial) != n_pars) {
       cli::cli_abort(
-        paste("Unexpected initial parameter length {length(pars)};",
-              "expected {length(model$parameters)}"),
-        arg = "initial")
+        c(paste("Unexpected length for vector 'initial'",
+                "(given {length(initial)}, expected {n_pars})"),
+          i = "I expected a vector with {n_pars} element{?s}, 1 per parameter"),
+        arg = "initial", call = call)
     }
+    initial <- matrix(initial, n_chains, n_pars, byrow = TRUE)
   }
+  initial
+}
 
-  density <- model$density(pars)
-  state <- list(pars = pars, density = density)
-  sampler$initialise(state, model, rng)
-
-  history_pars <- matrix(NA_real_, n_steps + 1, length(pars))
-  history_pars[1, ] <- pars
-  history_density <- rep(NA_real_, n_steps + 1)
-  history_density[[1]] <- density
-
-  for (i in seq_len(n_steps)) {
-    state <- sampler$step(state, model, rng)
-    history_pars[i + 1, ] <- state$pars
-    history_density[[i + 1]] <- state$density
-  }
 
-  ## Pop the parameter names on last
-  colnames(history_pars) <- model$parameters
+combine_chains <- function(res) {
+  stopifnot(is.null(res$chain))
+  chain <- rep(seq_along(res), lengths(lapply(res, "[[", "density")))
```

Comment:
Sorry, struggling to parse what this gets you. The lengths of each chain?

Reply:
a sequence 1, 1, 1, 1..., 2, 2, 2, 2, .... etc (1 * n_steps, 2 * n_steps, ..., n_chains * n_steps)

Reply:
Oh, so it tells you which chain each density and pars value comes from?

Reply:
yeah exactly that